Mathlib/CategoryTheory/Category/FreeGroupoid.lean

```diff
@@ -0,0 +1,220 @@
+-/
+import Mathlib.CategoryTheory.Groupoid.FreeGroupoid
+import Mathlib.CategoryTheory.Category.Grpd
+import Mathlib.CategoryTheory.Adjunction.Reflective
+import Mathlib.CategoryTheory.Localization.Predicate
+
+/-!
+# Free groupoid on a category
+
+This file defines the free groupoid on a category, the lifting of a functor to its unique
+extension as a functor from the free groupoid, and proves uniqueness of this extension.
+
+## Main results
+
+Given a type `C` and a category instance on `C`:
+
+- `CategoryTheory.Category.FreeGroupoid C`: the underlying type of the free groupoid on `C`.
+- `CategoryTheory.Category.FreeGroupoid.instGroupoid`: the `Groupoid` instance on `FreeGroupoid C`.
+- `CategoryTheory.Category.FreeGroupoid.lift`: the lifting of a functor `C ⥤ G` where `G` is a
+  groupoid, to a functor `CategoryTheory.Category.FreeGroupoid C ⥤ G`.
+- `CategoryTheory.Category.FreeGroupoid.lift_spec` and
+  `CategoryTheory.Category.FreeGroupoid.lift_unique`:
+  the proofs that, respectively, `CategoryTheory.Category.FreeGroupoid.lift` indeed is a lifting
+  and is the unique one.
+- `CategoryTheory.Category.Grpd.free`: the free functor from `Grpd` to `Cat`
+- `CategoryTheory.Category.Grpd.freeForgetAdjunction`: that `free` is left adjoint to
+  `Grpd.forgetToCat`.
+
+## Implementation notes
+
+The free groupoid on a category `C` is first defined by taking the free groupoid `G`
+on the underlying *quiver* of `C`. Then the free groupoid on the *category* `C` is defined as
+the quotient of `G` by the relation that makes the inclusion prefunctor `C ⥤q G` a functor.
+
+-/
+
+noncomputable section
+
+namespace CategoryTheory
+
+universe v u v₁ u₁ v₂ u₂
+
+namespace Category
+
+variable (C : Type u) [Category.{v} C]
+
+open Quiver in
+/-- The relation on the free groupoid on the underlying *quiver* of C that
+promotes the prefunctor `C ⥤q FreeGroupoid C` into a functor
+`C ⥤ Quotient (FreeGroupoid.homRel C)`. -/
+inductive FreeGroupoid.homRel : HomRel (FreeGroupoid C) where
+| map_id (X : C) : homRel ((FreeGroupoid.of C).map (𝟙 X)) (𝟙 ((FreeGroupoid.of C).obj X))
+| map_comp {X Y Z : C} (f : X ⟶ Y) (g : Y ⟶ Z) : homRel ((FreeGroupoid.of C).map (f ≫ g))
+  ((FreeGroupoid.of C).map f ≫ (FreeGroupoid.of C).map g)
+
+/-- The underlying type of the free groupoid on a category,
+defined by quotienting the free groupoid on the underlying quiver of `C`
+by the relation that promotes the prefunctor `C ⥤q FreeGroupoid C` into a functor
+`C ⥤ Quotient (FreeGroupoid.homRel C)`. -/
+protected def FreeGroupoid := Quotient (FreeGroupoid.homRel C)
+
+variable {C} in
+instance [Nonempty C] : Nonempty (Category.FreeGroupoid C) := by
+  have : Inhabited (Quiver.FreeGroupoid C) := by
+    inhabit Quiver.FreeGroupoid C
+    exact ⟨@default (Quiver.FreeGroupoid C) _⟩
+  have : Inhabited (Category.FreeGroupoid C) := inferInstanceAs (Inhabited <| Quotient _)
+  inhabit Category.FreeGroupoid C
+  exact ⟨@default (Category.FreeGroupoid C) _⟩
+
+variable {C} in
```

Comment:
The distinction between implicit/explicit variables do not matter very much for instance declarations. Here, I do not see the need to make it implicit:
```suggestion
```